src/eval_double.cpp

```diff
@@ -85,6 +89,135 @@ class EvalDoubleVisitor : public BaseVisitor<EvalDoubleVisitor> {
 
+    void bvisit(const Symbol &) {
+        throw std::runtime_error("Symbol cannot be evaluated as a double.");
+    };
+
+    void bvisit(const Log &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::log(tmp);
+    };
+
+    void bvisit(const Cot &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = 1/::tan(tmp);
+    };
+
+    void bvisit(const Csc &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = 1/::cos(tmp);
+    };
+
+    void bvisit(const Sec &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = 1/::sin(tmp);
+    };
+
+    void bvisit(const ASin &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::asin(tmp);
+    };
+
+    void bvisit(const ACos &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::acos(tmp);
+    };
+
+    void bvisit(const ASec &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::asin(1/tmp);
+    };
+
+    void bvisit(const ACsc &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::acos(1/tmp);
+    };
+
+    void bvisit(const ATan &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::atan(tmp);
+    };
+
+    void bvisit(const ACot &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::atan(1/tmp);
+    };
+
+    void bvisit(const ATan2 &x) {
+        double num = apply(*(x.get_num()));
+        double den = apply(*(x.get_den()));
+        result_ = ::atan2(num, den);
+    };
+
+    void bvisit(const Sinh &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::sinh(tmp);
+    };
+
+    void bvisit(const Cosh &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::cosh(tmp);
+    };
+
+    void bvisit(const Tanh &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::tanh(tmp);
+    };
+
+    void bvisit(const Coth &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = 1/::tanh(tmp);
+    };
+
+    void bvisit(const ASinh &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::asinh(tmp);
+    };
+
+    void bvisit(const ACosh &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::acosh(tmp);
+    };
+
+    void bvisit(const ATanh &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::atanh(tmp);
+    };
+
+    void bvisit(const ACoth &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::atanh(1/tmp);
+    };
+
+    void bvisit(const ASech &x) {
+        double tmp = apply(*(x.get_arg()));
+        result_ = ::acosh(1/tmp);
+    };
+
+    void bvisit(const Gamma &x) {
+        double tmp = apply(*(x.get_args()[0]));
+        result_ = ::tgamma(tmp);
+    };
+
+    void bvisit(const Constant &x) {
+        if (x.__eq__(*pi)) {
+            result_ = ::atan2(0, -1);
+        } else if (x.__eq__(*E)) {
+            result_ = ::exp(1);
+        } else {
+            throw std::runtime_error("Unknown constant.");
```

Comment:
I think it should be possible to evaluate every constant. So I would name the exception " This constant is not implemented yet.", or something like that.